src/main/java/org/mastodon/mamut/nearest/NearestObjectStatComputation.java

```diff
@@ -0,0 +1,185 @@
+import org.mastodon.mamut.nearest.NearestObjectStatModel.NearestObjectStatItem;
+import org.mastodon.mamut.nearest.NearestObjectStatModel.Value;
+import org.mastodon.spatial.SpatialIndex;
+import org.mastodon.tracking.linking.ProgressListeners;
+import org.mastodon.ui.ProgressListener;
+
+import gnu.trove.list.array.TDoubleArrayList;
+
+public class NearestObjectStatComputation
+{
+
+	public static final NearestObjectStatFeature compute( final Model model, final NearestObjectStatModel statModel, final int minTimepoint, final int maxTimepoint )
+	{
+		return compute( model, statModel, minTimepoint, maxTimepoint, ProgressListeners.voidLogger() );
+	}
+
+	public static NearestObjectStatFeature compute( final Model model, final NearestObjectStatModel statModel, final int minTimepoint, final int maxTimepoint, final ProgressListener progressListener )
+	{
+		final NearestObjectStatFeature feature = NearestObjectStatFeature.createFeature( statModel, model );
+
+		final ReadLock lock = model.getGraph().getLock().readLock();
+		lock.lock();
+		try
+		{
+			for ( int t = minTimepoint; t <= maxTimepoint; t++ )
+			{
+				progressListener.showProgress( t - minTimepoint, maxTimepoint - minTimepoint + 1 );
+				process( model, statModel, t, feature ).run();
+			}
+			return feature;
+		}
+		finally
+		{
+			progressListener.clearStatus();
+			lock.unlock();
+		}
+	}
+
+	private static final Runnable process( final Model model, final NearestObjectStatModel nosModel, final int t, final NearestObjectStatFeature feature )
+	{
+		return new Runnable()
+		{
+
+			@Override
+			public void run()
+			{
+				// Max number of neighbors we need to collect when we specify N.
+				int maxN = -1;
+				// Max distance we need to search for when we specify max dist.
+				double maxDistance = Double.NEGATIVE_INFINITY;
+
+				final List< ToDoubleBiFunction< Spot, Spot > > evaluators = new ArrayList<>( nosModel.size() );
+				for ( final NearestObjectStatItem item : nosModel )
+				{
+					final ToDoubleBiFunction< Spot, Spot > evaluator = item.eval( model.getFeatureModel() );
+					evaluators.add( evaluator );
+
+					switch ( item.collectBy )
+					{
+					case MAX_DISTANCE:
+					{
+						final double md = item.maxDistance;
+						if ( md > maxDistance )
+							maxDistance = md;
+						break;
+					}
+					case SPECIFY_N:
+					{
+						final int n = item.n;
+						if ( n > maxN )
+							maxN = n;
+						break;
+					}
+					default:
+						throw new IllegalArgumentException( "Unknown collection method: " + item.collectBy );
+					}
+				}
+				final RefList< Spot > list = RefCollections.createRefList( model.getGraph().vertices() );
+
+				// Spatial search.
+				final SpatialIndex< Spot > si = model.getSpatioTemporalIndex().getSpatialIndex( t );
+				final IncrementalNearestNeighborSearch< Spot > search = si.getIncrementalNearestNeighborSearch();
+
+				// Storage for numerical values fetched from neighbor.
+				final TDoubleArrayList arr = new TDoubleArrayList();
```

Comment:
Perhaps call this variable `values` or `neighborValues` instead? This may be easier to identify when reading this variable on other lines of the code